File: Example/objc/FlexiCollectionViewLayout/FlexiCollectionViewLayout/ViewController.m
//
//  ViewController.m
//  FlexiCollectionViewLayout
//
//  Created by Deepak Kumar on 11/30/16.
//  Copyright © 2016 Deepak Kumar. All rights reserved.
//

#import "ViewController.h"
#import "FlexiCollectionViewLayout-Swift.h"
#import "HKCell.h"

@interface ViewController ()<FlexiCollectionViewLayoutDelegate>
@property (nonatomic, strong) NSArray *hongKongPhotos;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.hongKongPhotos = @[[UIImage imageNamed: @"beach"], [UIImage imageNamed: @"boat"], [UIImage imageNamed: @"brucelee"], [UIImage imageNamed: @"dragonboat"], [UIImage imageNamed: @"icc"], [UIImage imageNamed: @"icclightshow"], [UIImage imageNamed: @"ifc"], [UIImage imageNamed: @"island"], [UIImage imageNamed: @"lantau"], [UIImage imageNamed: @"oceanpark"], [UIImage imageNamed: @"orange"], [UIImage imageNamed: @"panda"], [UIImage imageNamed: @"sunset"], [UIImage imageNamed: @"thepeak"], [UIImage imageNamed: @"tram"]];
    
    FlexiCollectionViewLayout *layout = [[FlexiCollectionViewLayout alloc] init];
    [self.collectionView setCollectionViewLayout:layout];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(NSInteger) numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 3;
}

-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return 28;
            break;
        case 1:
            return 20;
            break;
        case 2:
            return 24;
            break;
        default:
            return 24;
            break;
    }
}

- (ItemSizeAttributes *)collectionView:(UICollectionView *)collectionView layout:(FlexiCollectionViewLayout *)collectionViewLayout sizeForFlexiItemAt:(NSIndexPath *)indexPath {
    return [[ItemSizeAttributes alloc] initWithItemSize:CGSizeMake(120, 120) layoutSize:FlexiCellSizeRegular widthFactor:1 heightFactor:1];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    HKCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier: @"HKCellReuseID" forIndexPath:indexPath];
    [cell.imageView setImage: _hongKongPhotos[indexPath.item % _hongKongPhotos.count] ];
    [cell.textLabel setText: [NSString stringWithFormat:@"Image %ld", (long)indexPath.item]];
    return cell;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return 2;
}

- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section {
    return UIEdgeInsetsZero;
}



@end

File: Example/FlexiCollectionViewLayout/ViewController.swift
//
//  ViewController.swift
//  FlexiCollectionViewLayout
//
//  Created by Deepak Kumar on 11/13/16.
//  Copyright © 2016 Deepak Kumar. All rights reserved.
//

import UIKit
import FlexiCollectionViewLayout

internal let cellReuseIdentifier = "HKCellReuseID"

final class ViewController: UICollectionViewController, FlexiCollectionViewLayoutDelegate {
    
    private let hongKongPhotos = [UIImage(named: "beach"), UIImage(named: "boat"),UIImage(named: "brucelee"),UIImage(named: "dragonboat"),UIImage(named: "icc"),UIImage(named: "icclightshow"),UIImage(named: "ifc"),UIImage(named: "island"),UIImage(named: "lantau"),UIImage(named: "oceanpark"),UIImage(named: "orange"),UIImage(named: "panda"),UIImage(named: "sunset"),UIImage(named: "thepeak"),UIImage(named: "tram")]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let layout = FlexiCollectionViewLayout()
        collectionView?.collectionViewLayout = layout
        registerSupplementaryViews()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    private func registerSupplementaryViews() {
        collectionView?.register(UINib(nibName: "HeaderView", bundle: nil), forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, withReuseIdentifier: "HeaderReuseID")
        collectionView?.register(UINib(nibName: "FooterView", bundle: nil), forSupplementaryViewOfKind: UICollectionElementKindSectionFooter, withReuseIdentifier: "FooterReuseID")
    }

    // MARK: UICollectionViewDataSource
    
    override func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 3
    }
    
    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        switch section {
        case 0:
            return 28
        case 1:
            return 20
        default:
            return 24
        }
    }
    
    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellReuseIdentifier, for: indexPath) as! HKCell
        
        cell.imageView.image = hongKongPhotos[indexPath.item % hongKongPhotos.count]
        cell.infoLabel.text = "Image \(indexPath.row)"
        
        return cell
    }
    
    override func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        if kind == UICollectionElementKindSectionHeader {
            let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionElementKindSectionHeader, withReuseIdentifier: "HeaderReuseID", for: indexPath) as! HeaderView
            headerView.headerLabel.text = "Section \(indexPath.section)"
            return headerView
        } else if kind == UICollectionElementKindSectionFooter {
            let footerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionElementKindSectionFooter, withReuseIdentifier: "FooterReuseID", for: indexPath) as! FooterView
            footerView.footerLabel.text = "Footer \(indexPath.section)"
            return footerView
        }
        return UICollectionReusableView()
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: FlexiCollectionViewLayout, sizeForFlexiItemAt indexPath: IndexPath) -> ItemSizeAttributes {
        switch indexPath.row {
        case 0:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.large, widthFactor: 2, heightFactor: 2)
        case 1:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.regular, widthFactor: 1, heightFactor: 1)
        case 2:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.regular, widthFactor: 1, heightFactor: 1)
        case 3:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.large, widthFactor: 3, heightFactor: 1)
        case 4:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.large, widthFactor: 1, heightFactor: 3)
        case 5:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.regular, widthFactor: 1, heightFactor: 1)
        case 6:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.regular, widthFactor: 1, heightFactor: 1)
        case 7:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.large, widthFactor: 2, heightFactor: 1)
        case 8:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.regular, widthFactor: 1, heightFactor: 1)
        case 9:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.regular, widthFactor: 1, heightFactor: 1)
        case 10:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.large, widthFactor: 1, heightFactor: 2)
        case 11:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.large, widthFactor: 2, heightFactor: 2)
        default:
            return ItemSizeAttributes(itemSize: CGSize(width: 120, height: 100), layoutSize: FlexiCellSize.regular, widthFactor: 1, heightFactor: 1)
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return UIEdgeInsets(top: 0, left: 1, bottom: 0, right: 1)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return 2
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, heightForFooterInSection section: Int) -> CGFloat {
        return 45
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, heightForHeaderInSection section: Int) -> CGFloat {
        return 45
    }
    
    //MARK: Helpers
    
    fileprivate func sizeBasedOnSizeClass() -> CGFloat {
        if DeviceUtilsHelper().isDeviceiPhoneInPortrait(traitCollection) {
            return self.collectionView!.bounds.size.width - 10
        } else {
            return (self.collectionView!.bounds.size.width - 14)/3
        }
    }
}

